processors/firehose_to_es_processor/index.js

```diff
@@ -0,0 +1,104 @@
+  "logStream": "log_stream_name",
+  "subscriptionFilters": [
+    "subscription_filter_name"
+  ],
+  "logEvents": [
+    {
+      "id": "01234567890123456789012345678901234567890123456789012345",
+      "timestamp": 1510109208016,
+      "message": "log message 1"
+    },
+    {
+      "id": "01234567890123456789012345678901234567890123456789012345",
+      "timestamp": 1510109208017,
+      "message": "log message 2"
+    }
+    ...
+  ]
+}
+
+The data is additionally compressed with GZIP.
+
+The code below will:
+
+1) Gunzip the data
+2) Parse the json
+3) Set the result to ProcessingFailed for any record whose messageType is not DATA_MESSAGE, thus redirecting them to the
+   processing error output. Such records do not contain any log events. You can modify the code to set the result to
+   Dropped instead to get rid of these records completely.
+4) For records whose messageType is DATA_MESSAGE, extract the individual log events from the logEvents field, and pass
+   each one to the transformLogEvent method. You can modify the transformLogEvent method to perform custom
+   transformations on the log events.
+5) Concatenate the result from (4) together and set the result as the data of the record returned to Firehose. Note that
+   this step will not add any delimiters. Delimiters should be appended by the logic within the transformLogEvent
+   method.
+*/
+
+'use strict';
+
+const zlib = require('zlib');
+
+/**
+ * logEvent has this format:
+ *
+ * {
+ *   "id": "01234567890123456789012345678901234567890123456789012345",
+ *   "timestamp": 1510109208016,
+ *   "message": "log message 1"
+ * }
+ *
+ *
+ * The result must be returned in a Promise.
+ */
+function transformLogEvent(logEvent) {
+    const transformed_timestamp = new Date(1 * logEvent.timestamp).toISOString();
+    const transformed_payload = {
+        message: logEvent.message,
+        id: logEvent.id,
+        timestamp: transformed_timestamp
+    };
+    return Promise.resolve(transformed_payload);
+}
+
+exports.handler = (event, context, callback) => {
+    Promise.all(event.records.map(r => {
+        const buffer = new Buffer(r.data, 'base64');
+        const decompressed = zlib.gunzipSync(buffer);
+        const data = JSON.parse(decompressed);
+        if (data.messageType !== 'DATA_MESSAGE') {
+            return Promise.resolve({
+                recordId: r.recordId,
+                result: 'ProcessingFailed',
+            });
+        } else {
+            const promises = data.logEvents.map(transformLogEvent);
+            return Promise.all(promises).then(transformed => {
+                const payload = transformed[0];
+                const output = {
+                    "@id": payload.id,
+                    "@message": payload.message,
+                    "@owner": data.owner,
+                    "@log_group": data.logGroup,
+                    "@log_stream": data.logStream,
+                    "@timestamp": payload.timestamp
+                };
```

Comment:
This is good, though I would include the first layer of keys present in the ingested json object as well. Es will index these using dynamic mapping and speed up the queries. Without this, we lose a lot of benefit of the search index for non-trivial queries.

Historically, the reasons that only the first level of the JSON object keys were included are that 1. most queries will operate on the info in the higher levels of abstraction and 2. ES only accepts <= 1000 fields in an index. CloudTrail alone can generate more than this.

See L71 of the old `exporters/cwl_to_elk/index.js` for more.

Reply:
makes sense, I'll just need to account for cases where the message itself is not a json object (such as initiating/terminating log ouputs from lambda functions)

Reply:
true!

Reply:
✅ we tackled this